### PR TITLE
add user.referees/user.referrer and referral in api entity

### PR DESCRIPTION
**Pull request by hardywu**

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -25,6 +25,8 @@ class User < ApplicationRecord
   scope :active, -> { where(state: 'active') }
 
   before_validation :assign_uid
+  belongs_to :referrer, class_name: "User", optional: true, foreign_key: 'referral_id'
```
Comment (ec) [inline app/models/user.rb]: ruby prefer single quote syntax

File: app/models/user.rb [hunk 2]
```diff
@@ -25,6 +25,8 @@ class User < ApplicationRecord
...
+  has_many :referees, class_name: "User", foreign_key: 'referral_id'
```
Comment (ec) [inline app/models/user.rb]: the same

File: app/api/v2/entities/user_with_full_info.rb [hunk 3]
```diff
@@ -13,6 +13,7 @@ class UserWithFullInfo < Grape::Entity
         expose :level, documentation: { type: 'Integer' }
         expose :otp, documentation: { type: 'Boolean' }
         expose :state, documentation: { type: 'String' }
+        expose :referral_id, documentation: { type: 'Integer' }
```
Comment (ec) [inline app/api/v2/entities/user_with_full_info.rb:16]: changes in entity always cause error in related specs. fix the specs

**Review by ec — CHANGES_REQUESTED** [review]

**Comment by ec** [comment]
What is the purpose of returning referral_id on current stage?
What is the purpose of making a relations between referrals and referees? As barong doesnt have any in-app functionality for that, and doesn't have api to use it

**Comment by hardywu** [comment]
@ec  The Frontend need to know whether the referral goes through successfully, particularly if there is a reward notification for a referral.

**Comment by ec** [comment]
@hardywu 
But still, you dont add an api for relation logic. So this part of code is "dead" now, which shouldn't be here.

Then, about referral_id. Referral_id is an integer of user id.
ID comes from DB and we don't return it in API. (Only UID, which is identifier for frontend)

Still, with referral_id front still doesnt know anything about referrer

**Comment by hardywu** [comment]
how about using a Boolean attribute `referred: true/false`

**Comment by ec** [comment]
Well, i doesn't see any advantage of it. If frontend still dont know about WHO is referrer - what is the purpose of knowledge user is referred?

**Comment by hardywu** [comment]
it's not about WHO is referrer, but letting front know the referral go through. A user would never know if his referral_id is correctly typed on registration, nor can he/she notify his friend who referred him.

**Comment by ec** [comment]
Still, its not the functionality, that should be in the barong. At least, now.
I guess, we will come back to this question a bit later, when peatio or core frontend functionality will strongly require it

**Pull request closed without merging**